lib/Dialect/AIEX/Transforms/AIECtrlPacketToDma.cpp

```diff
@@ -82,56 +82,102 @@ struct AIECtrlPacketToDmaPass : AIECtrlPacketToDmaBase<AIECtrlPacketToDmaPass> {
-              auto rowToShimChanMap =
-                  getRowToShimChanMap(targetModel, WireBundle::DMA);
-              int shimChan = rowToShimChanMap[row];
-              shimDmaAllocName += "_chan" + std::to_string(shimChan);
-
-              StringRef metadata = builder.getStringAttr(shimDmaAllocName);
-              builder.create<NpuDmaMemcpyNdOp>(
-                  builder.getUnknownLoc(), newBlockArg, SmallVector<Value>{},
-                  SmallVector<Value>{}, SmallVector<Value>{},
-                  ArrayRef(staticOffsets), ArrayRef(staticSizes),
-                  ArrayRef(staticStrides), nullptr, metadata, 0, true, 0, 0, 0,
-                  0, 0, 0);
-
-              auto shimRow = builder.getI32IntegerAttr(0);
-              auto shimCol = builder.getI32IntegerAttr(col);
-              auto dir = builder.getI32IntegerAttr(1); // MM2S
-              auto chan = builder.getI32IntegerAttr(shimChan);
-              auto col_num = builder.getI32IntegerAttr(1);
-              auto row_num = builder.getI32IntegerAttr(1);
-              builder.create<AIEX::NpuSyncOp>(loc, shimCol, shimRow, dir, chan,
-                                              col_num, row_num);
-            });
+        auto ctrlPktOp = dyn_cast<NpuControlPacketOp>(&o);
+        if (!ctrlPktOp)
+          continue;
+
+        assert(batchIt != batches.end() &&
+               "Expected control packet to be in a batch");
+
+        int col = ctrlPktOp.getColumnFromAddr();
+        int row = ctrlPktOp.getRowFromAddr();
+
+        // Check if this is the first packet of a new batch,
+        // otherwise skip it.
+        if (batchIt->tileId != TileID{col, row})
+          continue;
+
+        // Emit the batched DMA operation for this (col, row) pair
+        const std::vector<int64_t> staticOffsets = {0, 0, 0,
+                                                    batchIt->startOffset};
+        const std::vector<int64_t> staticSizes = {1, 1, 1, batchIt->totalSize};
+        const std::vector<int64_t> staticStrides = {0, 0, 0, 1};
+
+        StringRef metadata = builder.getStringAttr(batchIt->shimDmaAllocName);
+        builder.create<NpuDmaMemcpyNdOp>(
+            builder.getUnknownLoc(), newBlockArg, SmallVector<Value>{},
+            SmallVector<Value>{}, SmallVector<Value>{}, ArrayRef(staticOffsets),
+            ArrayRef(staticSizes), ArrayRef(staticStrides), nullptr, metadata,
+            0, true, 0, 0, 0, 0, 0, 0);
+
+        auto shimRow = builder.getI32IntegerAttr(0);
+        auto shimCol = builder.getI32IntegerAttr(col);
+        auto dir = builder.getI32IntegerAttr(1); // MM2S
+        auto chan = builder.getI32IntegerAttr(batchIt->shimChan);
+        auto col_num = builder.getI32IntegerAttr(1);
+        auto row_num = builder.getI32IntegerAttr(1);
+        builder.create<AIEX::NpuSyncOp>(loc, shimCol, shimRow, dir, chan,
+                                        col_num, row_num);
+        ++batchIt;
+        if (batchIt == batches.end())
```

Comment:
ugh, this is a bit of slop I ignored. could just iterate over batches instead...